### DISCUSSION TITLE
Linking and search path

### DISCUSSION THREAD

**Discussion by thewh1teagle** [k2-fsa/sherpa-onnx#1202] [Q&A]
I'm confused with linking and search path while creating bindings to sherpa-onnx in [sherpa-rs/build.rs](https://github.com/thewh1teagle/sherpa-rs/blob/main/sys/build.rs)

Do I always need to link `kaldi-native-fbank-core`, `sherpa-onnx-core` etc? What if shared libs is on eg. with Cuda or DirectML?

As for `espeak-ng`, `piper_phonemize`. always link static? only when tts enabled?
I'm confuse with what to link and how.
Also regarding DLLs, they should always be available inside build/lib folder after compile?

Also, I like to link statically as far as possible. When it's not possible?

**Comment by csukuangfj** [comment]
You have to distinguish between a shared build and a static build of sherpa-onnx.

For a static build with TTS, please refer to
https://github.com/k2-fsa/sherpa-onnx/blob/35c1b4a7a9376c6bb80ac461e1b2169be563f908/cmake/sherpa-onnx-static.pc.in#L25

For a static build without TTS, please refer to
https://github.com/k2-fsa/sherpa-onnx/blob/35c1b4a7a9376c6bb80ac461e1b2169be563f908/cmake/sherpa-onnx-static-no-tts.pc.in#L25

For a shared build (regardless of with or without TTS), please refer to
https://github.com/k2-fsa/sherpa-onnx/blob/35c1b4a7a9376c6bb80ac461e1b2169be563f908/cmake/sherpa-onnx-shared.pc.in#L25


---

> What if shared libs is on eg. with Cuda or DirectML?

In that case, only shared lib should be used.


---

> Also regarding DLLs, they should always be available inside build/lib folder after compile?

Yes, as long as you run `make install`. Otherwise, the `onnxruntime` library won't be copied to `build/lib`.


---

> Also, I like to link statically as far as possible. When it's not possible?



Whe you use CUDA or directml, you have to use a shared lib of sherpa-onnx.

---

By the way, it you want to always use a static build of sherpa-onnx, you need to change CMakeLists.txt so that
it supports static building even if you use CUDA or directML.

The source code of sherpa-onnx can be built from the source, so it is possible to always use a static build.

**Reply by thewh1teagle** [reply]
Thank you! it really helps.
Now I use these 3 `.pc` files to know exactly what to link.
I understand that regarding the `DLLs` files they should be places under `lib` in the end and I don't need to search in another places.
Regarding when using directml or cuda - how can I know which lib names to link dynamic / static? I can't see `.pc` file of cuda.

Also, the search path with cuda/directml etc - should I set it to something like `build/_deps/onnxruntime-src/lib`?

**Reply by csukuangfj** [reply]
>  how can I know which lib names to link dynamic / static?

Please see what lib files are in `build/_deps/onnxruntime-src/lib`

---

> Also, the search path with cuda/directml etc - should I set it to something like build/_deps/onnxruntime-src/lib?

Yes, I think that is reasonable.
